src/main/java/com/yahoo/sketches/quantiles/MWMRHeapUpdateDoublesSketch.java

```diff
@@ -0,0 +1,852 @@
+
+			PropogationLock.set(1);
+
+			// copy to empty place in combained buffer
+			long bitPattern = ds.getBitPattern();
+			int startingLevel = ds.getNumberOfTreeLevels_() - 1;
+
+			long newN = ds.getN() + MWMRHeapUpdateDoublesSketch.numberOfLeaves(startingLevel + 1) * 2 * ds.k_;
+
+			// make sure there will be enough space (levels) for the propagation
+			final int spaceNeeded = DoublesUpdateImpl.getRequiredItemCapacity(k_, newN);
+			final int combBufItemCap = ds.getCombinedBufferItemCapacity();
+
+			if (spaceNeeded > combBufItemCap) {
+				// copies base buffer plus old levels, adds space for new level
+				growCombinedBuffer(combBufItemCap, spaceNeeded);
+			}
+
+			DoublesSketchAccessor tgtSketchBuf = DoublesSketchAccessor.wrap(ds, true);
+			final int endingLevel = Util.lowestZeroBitStartingAt(bitPattern, startingLevel);
+			tgtSketchBuf.setLevel(endingLevel);
+			tgtSketchBuf.putArray(TreeBuffer.getBuffer_(), 0, 0, ds.k_);
+
+			TreeBitPattern[0].set(0);
+			// TODO propagate
+
+			final DoublesSketchAccessor bbAccessor = DoublesSketchAccessor.wrap(ds, true);
+
+			for (int lvl = startingLevel; lvl < endingLevel; lvl++) {
+				assert (bitPattern & (1L << lvl)) > 0; // internal consistency check
+				final DoublesSketchAccessor currLevelBuf = tgtSketchBuf.copyAndSetLevel(lvl);
+				DoublesUpdateImpl.mergeTwoSizeKBuffers(currLevelBuf, // target level: lvl
+						tgtSketchBuf, // target level: endingLevel
+						bbAccessor);
+				DoublesUpdateImpl.zipSize2KBuffer(bbAccessor, tgtSketchBuf);
+			} // end of loop over lower levels
+
+			// update bit pattern with binary-arithmetic ripple carry
+			long newBitPattern = bitPattern + (1L << startingLevel);
+			assert newBitPattern == computeBitPattern(k_, newN); // internal consistency check
+			// assert newBitPattern == ds.getBitPattern() + 1 || newBitPattern = (1L <<
+			// startingLevel);
+
+			ds.putBitPattern(newBitPattern);
+			// baseBufferCount_ = 0;
+			ds.putN(newN);
+			PropogationLock.set(0);
+
+		}
+
+	}
+
+	public class BackgroundBaseTreeWriter extends BackgroundWriter { // Important: only one writer can generate this job
+																		// per base node.
+
+		private int nodeNumber_;
+
+		public BackgroundBaseTreeWriter(String type, int nodesNumber, MWMRHeapUpdateDoublesSketch ds) {
+			super(type, ds);
+			nodeNumber_ = nodesNumber;
+		}
+
+		@Override
+		public void run() {
+
+//			setAffinity();
+
+			MWMRHeapUpdateDoublesSketch ds = getSketch();
+			DoublesArrayAccessor TreeBaseBuffer = ds.getTreeBaseBuffer_();
+			AtomicInteger[] TreeBaseBitPattern = ds.getTreeBaseBitPattern_();
+			DoublesArrayAccessor TreeBuffer = ds.getTreeBuffer_();
+			AtomicInteger[] TreeBitPattern = ds.getTreeBitPattern_();
+			AtomicInteger[] TreeReadyBitPattern = ds.getTreeReadyBitPattern_();
+
+			assert (TreeBaseBitPattern[nodeNumber_ - 1].get() == 1);
+			int NumberOfTreeInerNodes = (int) Math.pow(2, ds.getNumberOfTreeLevels_() - 1) - 1;
```

Comment:
can write the base buffer (2K) as the next level in the tree and then easy to find the parent in the tree and can use just one bit pattern for the tree